14_ComputeReduce/main.cpp

```diff
@@ -0,0 +1,281 @@
+		system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
+
+	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		if (!device_base_t::onAppInitialized(std::move(system)))
+			return false;
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+			return false;
+
+		computeQueue = getComputeQueue();
+
+		// Create (an almost) 128MB input buffer
+		constexpr auto in_size = 128u << 10u;
+		constexpr auto in_count = in_size / sizeof(uint32_t) - 23u;
+
+		m_logger->log("Input element count: %d", ILogger::ELL_PERFORMANCE, in_count);
+
+		inputData = new uint32_t[in_count];
+		{
+			std::random_device random_device;
+			std::mt19937 generator(random_device());
+			std::uniform_int_distribution<uint32_t> distribution(0u, ~0u);
+			for (auto i = 0u; i < in_count; i++)
+				inputData[i] = distribution(generator) % 100000;
+		}
+		auto minSSBOAlign = m_physicalDevice->getLimits().minSSBOAlignment;
+		constexpr auto begin = in_count / 4 + 118;
+		assert(((begin * sizeof(uint32_t)) & (minSSBOAlign - 1u)) == 0u);
+		constexpr auto end = in_count * 3 / 4 - 78;
+		assert(((end * sizeof(uint32_t)) & (minSSBOAlign - 1u)) == 0u);
+		constexpr auto elementCount = end - begin;
+
+		// Set Semaphores to control GPU synchronization
+		core::smart_refctd_ptr<ISemaphore> semaphore = m_device->createSemaphore(0);
+		IQueue::SSubmitInfo::SSemaphoreInfo semInfo[1] = { {
+			.semaphore = semaphore.get(),
+			.value = 1,
+			.stageMask = PIPELINE_STAGE_FLAGS::ALL_TRANSFER_BITS
+		} };
+
+		smart_refctd_ptr<IGPUBuffer> gpuinputDataBuffer;
+		{
+			IGPUBuffer::SCreationParams inputDataBufferCreationParams = {};
+			inputDataBufferCreationParams.size = sizeof(uint32_t) * in_count; // TODO Declare the element data type in the shader?
+			inputDataBufferCreationParams.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_SRC_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+			auto temp = m_utils->createFilledDeviceLocalBufferOnDedMem(
+				SIntendedSubmitInfo{ .queue = getTransferUpQueue() },
+				std::move(inputDataBufferCreationParams),
+				inputData,
+				{ semInfo, 1 }
+			);
+
+			const ISemaphore::SWaitInfo semWaitInfo[] = { {
+				.semaphore = semaphore.get(),
+				.value = 1
+			} };
+			if (m_device->blockForSemaphores(semWaitInfo) != ISemaphore::WAIT_RESULT::SUCCESS) {
+				m_logger->log("Blocking for operation semaphore failed during input data buffer creation", ILogger::ELL_ERROR);
+				return false;
+			}
+			gpuinputDataBuffer = *temp.get();
+		}
+		SBufferRange<IGPUBuffer> in_gpu_range = { begin * sizeof(uint32_t), elementCount * sizeof(uint32_t), gpuinputDataBuffer };
+
+		auto reducer = m_utils->getDefaultReducer();
+
+		CArithmeticOps::DefaultPushConstants reduce_push_constants;
+		CArithmeticOps::DispatchInfo reduce_dispatch_info;
+		reducer->buildParameters(elementCount, reduce_push_constants, reduce_dispatch_info);
```

Comment:
I think I've left a comment in the "mother PR" that this should be a constructor of parameters or something

Kinda silly to have to use a factory to do some basic arithmetic